python/glow/gwas/tests/test_approx_firth.py

```diff
@@ -0,0 +1,138 @@
+import glow.gwas.log_reg as lr
+import glow.gwas.approx_firth as af
+import glow.functions as fx
+import pandas as pd
+from pandas.testing import assert_frame_equal
+from pyspark.ml.linalg import DenseMatrix
+from pyspark.sql import Row
+import numpy as np
+import pytest
+
+
+default_phenotypes = [1, 0, 0, 1, 1, 1, 1, 0]
+default_genotypes = [0, 0, 1, 2, 2, 1, 1, 1]
+default_covariates = [[1, 1, 1, 1, 1, 1, 1, 1], [2, 3, 2, 3, 2, 3, 2, 3], [-1, -2, -3, -1, -2, -3, -1, -2]]
+default_offset = [0.1, 0.2, 0.3, 0.4, 0.4, 0.3, 0.2, 0.1]
+
+
+def get_golden_firth_fit(
+        spark,
+        phenotypes=default_phenotypes,
+        genotypes=default_genotypes,
+        covariates=default_covariates,
+        offset=default_offset):
+
+    covariate_matrix = DenseMatrix(
+        numRows=len(covariates[0]),
+        numCols=len(covariates),
+        values=list(np.ravel(covariates))
+    )
+    df = spark.createDataFrame([Row(
+        genotypes=genotypes,
+        phenotypes=phenotypes,
+        covariates=covariate_matrix,
+        offset=offset
+    )])
+    return df.select(fx.expand_struct(
+        fx.logistic_regression_gwas('genotypes', 'phenotypes', 'covariates', 'Firth', 'offset')
+    ))
+
+
+def _compare_full_firth_beta(
+        spark,
+        phenotypes=default_phenotypes,
+        genotypes=default_genotypes,
+        covariates=default_covariates,
+        offset=default_offset):
+
+    golden_firth_fit = get_golden_firth_fit(spark, phenotypes, genotypes, covariates, offset)
+
+    test_firth_fit = af._fit_firth(
+        beta_init=np.zeros(len(covariates) + 1),
+        X=np.column_stack([np.array(covariates).T, genotypes]),
+        y=np.array(phenotypes),
+        offset=np.array(offset),
+    )
+    golden_firth_beta = golden_firth_fit.head().beta
+    test_firth_beta = test_firth_fit.beta[-1]
+    assert np.allclose(golden_firth_beta, test_firth_beta)
+
+
+def test_full_firth(spark):
```

Comment:
For these tests, I'd actually prefer to just have literal golden values that can be cross referenced against R. I would prefer to not have a long term dependency on the SQL regression functions.